File: backend/src/models/UserInteraction.js
const mongoose = require("mongoose");

const userInteractionSchema = new mongoose.Schema({
  userId: { type: mongoose.Schema.Types.ObjectId, ref: "User", required: true },
  policyId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: "Policy",
    required: true,
  },
  type: {
    type: String,
    enum: ["view", "vote", "comment"],
    required: true,
  },
  createdAt: { type: Date, default: Date.now },
});

module.exports = mongoose.model("UserInteraction", userInteractionSchema);

File: mobile/civic_engagement_app/lib/src/features/citizen/domain/entities/user_interaction.dart
import 'package:equatable/equatable.dart';

/// Types of user interactions with policies
enum InteractionType {
  view,
  vote,
  comment;

  String toJson() => name;

  static InteractionType fromJson(String json) {
    return InteractionType.values.firstWhere(
      (type) => type.name == json,
      orElse: () => InteractionType.view,
    );
  }
}

/// Represents a user interaction with a policy
class UserInteraction extends Equatable {
  final String policyId;
  final InteractionType type;

  const UserInteraction({
    required this.policyId,
    required this.type,
  });

  Map<String, dynamic> toJson() => {
        'policyId': policyId,
        'type': type.toJson(),
      };

  @override
  List<Object?> get props => [policyId, type];
}
